crates/control/src/role_assignment.rs

```diff
@@ -662,114 +687,66 @@ fn pick_role_with_penalties(
-        && role_assignment[PlayerNumber::Four].is_none()
-        && penalties[PlayerNumber::Four].is_none()
+        && needs_assignment(PlayerNumber::Four, penalties, &role_assignment)
     {
         role_assignment[PlayerNumber::Four] = Some(Role::ReplacementKeeper);
         unassigned_robots -= 1;
     } else if !keeper_or_replacement_keeper_found
-        && role_assignment[PlayerNumber::Five].is_none()
-        && penalties[PlayerNumber::Five].is_none()
+        && needs_assignment(PlayerNumber::Five, penalties, &role_assignment)
     {
         role_assignment[PlayerNumber::Five] = Some(Role::ReplacementKeeper);
         unassigned_robots -= 1;
     }
 
-    if unassigned_robots > 0
-        && role_assignment[PlayerNumber::Two].is_none()
-        && penalties[PlayerNumber::Two].is_none()
-    {
-        role_assignment[PlayerNumber::Two] = Some(Role::DefenderRight);
-        unassigned_robots -= 1;
-    } else if unassigned_robots > 0
-        && role_assignment[PlayerNumber::Three].is_none()
-        && penalties[PlayerNumber::Three].is_none()
-    {
-        role_assignment[PlayerNumber::Three] = Some(Role::DefenderRight);
-        unassigned_robots -= 1;
-    } else if unassigned_robots > 0
-        && role_assignment[PlayerNumber::Four].is_none()
-        && penalties[PlayerNumber::Four].is_none()
-    {
-        role_assignment[PlayerNumber::Four] = Some(Role::DefenderRight);
-        unassigned_robots -= 1;
-    } else if unassigned_robots > 0
-        && role_assignment[PlayerNumber::Five].is_none()
-        && penalties[PlayerNumber::Five].is_none()
-    {
-        role_assignment[PlayerNumber::Five] = Some(Role::DefenderRight);
-        unassigned_robots -= 1;
-    }
+    // assign optional roles
+    let mut optional_role_index = 0;
 
-    if unassigned_robots > 0
-        && role_assignment[PlayerNumber::Two].is_none()
-        && penalties[PlayerNumber::Two].is_none()
-    {
-        role_assignment[PlayerNumber::Two] = Some(Role::StrikerSupporter);
-        unassigned_robots -= 1;
-    } else if unassigned_robots > 0
-        && role_assignment[PlayerNumber::Three].is_none()
-        && penalties[PlayerNumber::Three].is_none()
-    {
-        role_assignment[PlayerNumber::Three] = Some(Role::StrikerSupporter);
-        unassigned_robots -= 1;
-    } else if unassigned_robots > 0
-        && role_assignment[PlayerNumber::Four].is_none()
-        && penalties[PlayerNumber::Four].is_none()
-    {
-        role_assignment[PlayerNumber::Four] = Some(Role::StrikerSupporter);
-        unassigned_robots -= 1;
-    } else if unassigned_robots > 0
-        && role_assignment[PlayerNumber::Five].is_none()
-        && penalties[PlayerNumber::Five].is_none()
-    {
-        role_assignment[PlayerNumber::Five] = Some(Role::StrikerSupporter);
-        unassigned_robots -= 1;
+    if unassigned_robots > optional_roles.len() {
+        panic!("role_assignment not possible: not enough optional_roles available")
```

Comment:
Do you really want to panic here? Consider failing more softly instead of crashing the framework

Reply:
Removed in favor of resorting to `Role::default()` if things fail.

Reply:
There is an ongoing discussion on logging at #91 . I don't know what to conclude, but in my opinion this would be a case, where one should place a warning log?

Reply:
Yes. I think if the role-assignment results in a Striker-role, without the Player-number matching that of the Striker, we could/should do "something" ... logging, panic, warning, error ... all fine with me.

I'd let the ongoing discussion conclude and add the desired choice in the 7vs7 PR.

Reply:
For now, I suggest resolving this.
Tests in Behavior Simulator look good.
If you do not have more change requests, this is ready for merge.